File: backend/automations/src/main/java/com/dashboard/automations/service/N8nService.java
package com.dashboard.automations.service;

import com.dashboard.automations.dto.N8nSettingsRequest;
import com.dashboard.automations.model.User;
import com.dashboard.automations.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestTemplate;

import java.util.Map;

@Service
@RequiredArgsConstructor
public class N8nService {

    private final UserRepository userRepository;
    private final RestTemplate restTemplate;

    @Transactional
    public void saveSettings(String email, N8nSettingsRequest request) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
        String url = request.getN8nUrl().endsWith("/")
            ? request.getN8nUrl().substring(0, request.getN8nUrl().length() - 1)
            : request.getN8nUrl();
        user.setN8nUrl(url);
        user.setN8nApiKey(request.getN8nApiKey());
    }

    public Object getWorkflows(String email) {
        User user = getUserWithN8n(email);
        return proxyGet(user, "/api/v1/workflows");
    }

    public Object getWorkflow(String email, String workflowId) {
        User user = getUserWithN8n(email);
        return proxyGet(user, "/api/v1/workflows/" + workflowId);
    }

    public Object getExecutions(String email, String workflowId) {
        User user = getUserWithN8n(email);
        return proxyGet(user, "/api/v1/executions?workflowId=" + workflowId + "&limit=25");
    }

    public Object activateWorkflow(String email, String workflowId) {
        User user = getUserWithN8n(email);
        return proxyPost(user, "/api/v1/workflows/" + workflowId + "/activate", null);
    }

    public Object deactivateWorkflow(String email, String workflowId) {
        User user = getUserWithN8n(email);
        return proxyPost(user, "/api/v1/workflows/" + workflowId + "/deactivate", null);
    }

    public Object runWorkflow(String email, String workflowId, Map<String, Object> body) {
        User user = getUserWithN8n(email);
        return proxyPost(user, "/api/v1/workflows/" + workflowId + "/run", body);
    }

    private Object proxyGet(User user, String path) {
        HttpHeaders headers = buildHeaders(user.getN8nApiKey());
        ResponseEntity<Object> response = restTemplate.exchange(
                user.getN8nUrl() + path,
                HttpMethod.GET,
                new HttpEntity<>(headers),
                Object.class
        );
        return response.getBody();
    }

    private Object proxyPost(User user, String path, Object body) {
        HttpHeaders headers = buildHeaders(user.getN8nApiKey());
        ResponseEntity<Object> response = restTemplate.exchange(
                user.getN8nUrl() + path,
                HttpMethod.POST,
                new HttpEntity<>(body, headers),
                Object.class
        );
        return response.getBody();
    }

    private HttpHeaders buildHeaders(String apiKey) {
        HttpHeaders headers = new HttpHeaders();
        headers.set("X-N8N-API-KEY", apiKey);
        headers.set("Content-Type", "application/json");
        return headers;
    }

    private User getUserWithN8n(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
        if (user.getN8nUrl() == null || user.getN8nApiKey() == null) {
            throw new RuntimeException("n8n not configured");
        }
        return user;
    }
}

File: frontend/src/services/n8n.service.ts
import api from '@/lib/api'
import type { N8nWorkflowsResponse, N8nExecutionsResponse, N8nWorkflow } from '@/types'

export const n8nService = {
  saveSettings: (data: { n8nUrl: string; n8nApiKey: string }) =>
    api.put('/api/n8n/settings', data),

  getWorkflows: () =>
    api.get<N8nWorkflowsResponse>('/api/n8n/workflows').then((r) => r.data),

  getAllExecutions: () =>
    api.get<N8nExecutionsResponse>('/api/n8n/executions').then((r) => r.data),

  getWorkflow: (id: string) =>
    api.get<N8nWorkflow>(`/api/n8n/workflows/${id}`).then((r) => r.data as N8nWorkflow),

  getExecutions: (workflowId: string) =>
    api.get<N8nExecutionsResponse>(`/api/n8n/workflows/${workflowId}/executions`).then((r) => r.data),

  activate: (workflowId: string) =>
    api.post(`/api/n8n/workflows/${workflowId}/activate`).then((r) => r.data),

  deactivate: (workflowId: string) =>
    api.post(`/api/n8n/workflows/${workflowId}/deactivate`).then((r) => r.data),

  run: (workflowId: string, body?: Record<string, unknown>) =>
    api.post(`/api/n8n/workflows/${workflowId}/run`, body ?? {}).then((r) => r.data),

  updateWorkflow: (workflowId: string, body: Record<string, unknown>) =>
    api.put<N8nWorkflow>(`/api/n8n/workflows/${workflowId}`, body).then((r) => r.data),

  createWorkflow: (body: Record<string, unknown>) =>
    api.post<N8nWorkflow>('/api/n8n/workflows', body).then((r) => r.data),
}
